feat(aio): hide sidenav when current doc is not in sidenav menu

wardbell opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```
Igor requested this feature. He only wants the sidenav to appear when looking at guide pages. Marketing and API pages should use the whole screen.

Shows the left sidenav when the current doc is in the left nav (e.g., core or tutorial pages).
Hides the nav when the current doc is _not_ in the left nav (e.g. a marketing page or the API page).

To make this work, I changed `NavigationService` so that ONLY sidenav nodes can be in `selectedNodes`.  There is no good reason to have top menu nodes among `selectedNodes` because only `nav-items` use that feature.

This change makes the logic simple: if the current doc has "selected nodes", it's in the left sidenav; otherwise it is not.

The show/hide sidenav is now the intersection of selected node and screen width.

wardbell committed:
feat(aio): hide sidenav when current doc is not in sidenav menu

petebacondarwin commented on aio/e2e/app.e2e-spec.ts:
```diff
@@ -22,12 +22,13 @@ describe('site App', function() {
-    expect(heading.getText()).toMatch(/tutorial/i);
-    heading.click();
-    page.getLink('tutorial/').click();
+    // Show the menu; the tutorial section should be fully open from previous visit
```
+1

petebacondarwin commented on aio/src/app/app.component.html:
```diff
@@ -1,13 +1,13 @@
 <md-toolbar color="primary" class="app-toolbar">
-  <button *ngIf="isHamburgerVisible" class="hamburger" md-button (click)="sidenav.toggle()"><md-icon>menu</md-icon></button>
+  <button *ngIf="isHamburgerVisible" class="hamburger" md-button (click)="sidenav.toggle()" title="Docs menu"><md-icon>menu</md-icon></button>
```
"Docs menu" seems like a rather informal description.

wardbell replied:
What do you suggest?

petebacondarwin replied:
Perhaps "Documentation menu"?

petebacondarwin commented on aio/src/app/navigation/navigation.service.spec.ts:
```diff
@@ -120,7 +120,7 @@ describe('NavigationService', () => {
+    it('should be an empty array if no side navigation node matches the current location', () => {
       location.urlSubject.next('g');
       expect(currentNodes).toEqual([]);
     });
```
I think it is worth putting in a test that show that the `selectedNodes` doesn't emit if the urlSubject maps to a value that is a node but not in the side nav. Currently the tests only check that it doesn't emit if the url maps to something that is not in any node tree.

wardbell replied:
Done

petebacondarwin commented on aio/src/app/navigation/navigation.service.ts:
```diff
@@ -118,11 +117,11 @@ export class NavigationService {
+  private computeUrlToSideNavNodesMap(navigation: NavigationViews) {
     const navMap: NavigationMap = {};
-    Object.keys(navigation).forEach(key => navigation[key].forEach(node => walkNodes(node)));
+    navigation['SideNav'].forEach(node => walkNodes(node));
```
I would be tempted to pull this magic literal out into a const, to make it easier to find in the future.

petebacondarwin commented on aio/src/app/app.component.html:
```diff
@@ -1,13 +1,13 @@
 <md-sidenav-container class="sidenav-container">
 
-  <md-sidenav #sidenav class="sidenav" [opened]="isSideBySide" [mode] = "isSideBySide ? 'side' : 'over'">
+  <md-sidenav #sidenav class="sidenav" [opened]="isSideBySide && isSideNavNode" [mode]="isSideBySide ? 'side' : 'over'">
```
looks good - good spot on the inconsistent whitepace

petebacondarwin commented on aio/src/app/app.component.spec.ts:
```diff
@@ -67,6 +72,35 @@ describe('AppComponent', () => {
+
+    it('should have sidenav open when doc in the sidenav (guide/pipes)', fakeAsync(() => {
+      locationService.urlSubject.next('guide/pipes');
+      tick();
```
there is no need for `fakeAsync` nor `tick()` in these tests. They are both synchronous

wardbell replied:
Ah ... that wasn't true when before I faked `Http`. They were async at first (with `done()`) and then, when I faked `Http`, I didn't go all the way. Thanks for pointing this out.

petebacondarwin commented on aio/src/app/app.component.ts:
```diff
@@ -19,6 +19,7 @@ export class AppComponent implements OnInit {
   readonly homeImageUrl = 'assets/images/logos/standard/logo-nav.png';
 
   isHamburgerVisible = true; // always ... for now
+  isSideNavNode = false;
```
Given that we are hiding the logic about this being decided by the sidenav in the `NavigationService`, we should not be leaking that into the `AppComponent`. All the `AppComponent` cares about is whether there are any values in the `selectedNodes` or not.

So I would rather this was just called, `hasSelectedNodes` or something like that.

wardbell replied:
You're calling for a variable name change? `isSideNavNode` v. `hasSelectedNodes`?

I disagree.

1) The `AppComponent` needs `selectedNodes` anyway so it can pass them along to `NavMenuComponent`

2) Semantically, the `AppComponent`'s decides whether to display the sidenav based on whether the current doc is a side node; the fact that `selectedNodes` is involved in that determination is an implementation detail. Therefore, `isSideNavNode` is the better variable name.

3) The real leakage here is that the `AppComponent` is telling the `NavMenu` about nodes at all. The `NavMenuComponent` should get that information for itself.  I'm not a fan of

```
    <aio-nav-menu [nodes]="(navigationViews | async)?.SideNav" [selectedNodes]="selectedNodes | async"></aio-nav-menu>
```
which could have been (should be?)
```
    <aio-nav-menu (isSideNavDoc)="..."></aio-nav-menu>
```

4) But that involves a significant refactoring of responsibilities from the `AppComponent` to the `NavMenuComponent`.  IIRC, that's the way I had it in the beginning but you refactored it into the `AppComponent`.  I'd be happy to revisit that decision **but not in this PR**.

Therefore, I strongly urge that we keep this "as is" for this PR and consider refactoring the responsibility into the `NavMenuComponent` in a future PR.

petebacondarwin replied:
I agree not to rename it. Instead I prefer the suggestion below.

petebacondarwin replied:
Actually if we don't want the AppComponent to know about the implementation of whether the navigation is in the sidenav then we should move that computation into the NavigationService too and not do it in the AppComponent subscription

petebacondarwin replied:
Also it is not wrong to pass the nodes into the aio-nav-menu since that would prevent it from being reused to display other menus in other parts of the page.

petebacondarwin commented on aio/src/app/app.component.ts:
```diff
@@ -59,6 +60,9 @@ export class AppComponent implements OnInit {
     this.locationService.currentUrl.subscribe(url => this.autoScroll.scroll(this.docViewer.nativeElement.offsetParent));
+
+    // The current doc is in side nav if there are selected nodes
+    this.selectedNodes.subscribe(nodes => this.isSideNavNode = !!nodes.length );
```
We are now creating two subscriptions on the `selectedNodes` observable now: one here and one using the `async` pipe in the template.

Let's just have one subscription and write the emitted `selectedNodes` onto the component to be used in the template:

```ts
this.selectedNodes.subscribe(nodes => this.selectedNodes = nodes);
```

Then we don't need the `isSideNav` property at all because in our template we can do:

```html
<md-sidenav ...  [opened]="isSideBySide && selectedNodes.length" ...>
```

and

```html
<aio-nav-menu ... [selectedNodes]="selectedNodes"></aio-nav-menu>
````

What do you think? It would reduce the complexity of the component and its template.

wardbell replied:
Yes. But before we go there, see my previous comment about moving all of the logic down to the `NavMenuComponent`, which obviates this issue.

To reinforce, we also have TWO subscriptions to `navigationViews` in the HTML template and I didn't love that either.

petebacondarwin reviewed (changes requested):
A bunch of minor comments and two significant ones:

* remove `fakeAsync` from the tests
* consider refactoring the `selectedNodes` subscription.

Oh and it will need rebasing after #15411

petebacondarwin commented:
I do not believe it is a good idea to make the `NavMenuComponent` "smart". As it stands, it is reusable, which actually might come in useful for displaying the `TopBar ` navigation nodes in the side nav when the window is narrow.

Given that we now have up to 3 subscriptions to the `navigationNodes` observable, in the template, and also 1 template-based and now one code-based subscription for the `selectedNodes` observable, I think we should move the subscription into the component code and just expose the resulting values to the template.

But that is all refactoring that is not that relevant to this PR.

For this PR, I think we could debate the design choices for hours, so in the interests of moving this forward (and that I have previously declared that we should not block PRs on minor design arguments). I am happy to approve this PR with the design as it is.

There is one issue that does need to be fixed, though, which I just noticed. That is that, the subscription that is created in `ngOnInit` needs to be destroyed along with the component.

wardbell committed:
refactor(aio): app.component delegates to nav-menu & top-menu

Plus small cleanups, e.g., scroll, subscription consolidation, search template inlined.

mary-poppins commented:
The angular.io preview for 3f739fa is available [here][1].

[1]: https://pr15483-3f739fa463c0f9f019962632a89efd62d534753d.ngbuilds.io/

wardbell commented:
Aaagh!   

I got so annoyed with the complexity of `AppComponent` - it injects too much and knows too much about how the parts work - that I refactored into an additional commit on this PR.

Only to return here and find that @petebacondarwin LIKES that mess and doesn't want to see the nav menu and top menu take care of their own business (clearly I am editorializing).

In MY world, `AppComponent` is supposed to be as close to dead simple as possible. So I VERY much like what I did in that second commit.

Re: Pete's concern about overly `nav-menu`
1) YAGNI to say it should be reusable but even so ... it is no less reusable now. The fact that it injects the service it requires is reasonable. Why should expect some outer component to supply those nodes?

2) The same goes for the top menu.

3) Nothing stops us from reusing the top menu within the nav-menu should that become desirable. Could do that today if we wanted to in < 15 minutes. The top menu is also self sufficient.

4) Look at how much simpler the `AppComponent` is now?  It doesn't know all of the details necessary to drive the navigation components (top and side). Instead, it manages the central doc viewer and search.

5) I'd argue that it knows too much about that as well. A future refactoring IMO.

I realize that this is a dog fight between the smart/dumb architecture and an object-oriented view that emphasizes encapsulation and single-purpose classes.  IMO - and in the opinion of the guy who invented the smart/dumb distinction, the smart/dumb _is a mistake that he wishes he could take back_.

I'm going to argue strenuously for the refactor in the second commit.

wardbell commented:
@petebacondarwin I removed the `fakeAsync` in the FIRST commit (and force pushed it). I addressed the `selectedNodes` issue both in comments above and in the SECOND commit.

mary-poppins commented:
The angular.io preview for 40c04cc is available [here][1].

[1]: https://pr15483-40c04cc763886036b0b48d58ac9d02972f7254ef.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5ebb5dd is available [here][1].

[1]: https://pr15483-5ebb5dd9ff292ea77a0229ee7a892523a8fef79e.ngbuilds.io/

wardbell committed:
refactor(aio): AppComponent’s footer -> FooterComponent; gaService calls -> LocationService

wardbell commented:
In the same vein, because I had an hour before bed, I added 3rd commit that refactors `AppComponent` footer into `FooterComponent` and GA tracking to `LocationService`. Updated tests and added a couple more.

mary-poppins commented:
The angular.io preview for 7e3c526 is available [here][1].

[1]: https://pr15483-7e3c526a29e0565311e384fcb09060e71b6b5ead.ngbuilds.io/

petebacondarwin commented:
See #15492 for why `ViewMenuComponent` should not be "smart".

petebacondarwin commented:
+1 for the footer component

wardbell committed:
refactor(aio): isSideNavDoc not an emitter + other cleaning

mary-poppins commented:
The angular.io preview for b1c0096 is available [here][1].

[1]: https://pr15483-b1c00960dd891b865b9760543b62240070ccaa6e.ngbuilds.io/

mary-poppins commented:
The angular.io preview for b6aaaac is available [here][1].

[1]: https://pr15483-b6aaaac03fddb501b3fe33d31bc06e957413e3a8.ngbuilds.io/

wardbell committed:
refactor(aio): selectedNode -> currentNode for both side & top

mary-poppins commented:
The angular.io preview for 8012e85 is available [here][1].

[1]: https://pr15483-8012e85278d60b6433ef1711f711ac275f2245ab.ngbuilds.io/

wardbell commented:
Closed in favor of #15570
